### PR TITLE
Prefetch CSS only once

### PR DESCRIPTION
**Pull request by Jelenkee**
## Changes

When multiple links where prefetched the same CSS files where fetched multiple times as well.
This PR changes the behaviour so that a CSS file is not fetched more than once.

### CODE REVIEW

**Comment by natemoo-re** [comment]
This seems like a totally reasonable change! I'll ping @tony-sull to take a look at this as well. Any idea why tests are failing?

**Review by matthewp — APPROVED** [review]
lgtm

**Comment by Jelenkee** [comment]
Sorry, I am not able to run the tests on my local machine.
I use the ubuntu docker image to run the tests but I get some errors which are not related to my changes (I assume).
```
@astrojs/prefetch:test: Running 4 tests using 1 worker
@astrojs/prefetch:test: 
@astrojs/prefetch:test:   ✘  1 [Chrome Stable] › test/basic-prefetch.test.js:19:4 › Basic prefetch › dev › prefetches rel="prefetch" links › skips /admin (147ms)
@astrojs/prefetch:test: [vite:reporter] process.stdout.clearLine is not a function
@astrojs/prefetch:test:   ✘  2 [Chrome Stable] › test/basic-prefetch.test.js:55:4 › Basic prefetch › build › prefetches rel="prefetch" links › skips /admin (2ms)
@astrojs/prefetch:test:   ✘  3 [Chrome Stable] › test/custom-selectors.test.js:27:4 › Custom prefetch selectors › dev › prefetches links by custom selector › only prefetches /contact (155ms)
@astrojs/prefetch:test: [vite:reporter] process.stdout.clearLine is not a function
@astrojs/prefetch:test:   ✘  4 [Chrome Stable] › test/custom-selectors.test.js:60:4 › Custom prefetch selectors › build › prefetches links by custom selector › only prefetches /contact (2ms)
@astrojs/prefetch:test: 
@astrojs/prefetch:test: 
@astrojs/prefetch:test:   1) [Chrome Stable] › test/basic-prefetch.test.js:19:4 › Basic prefetch › dev › prefetches rel="prefetch" links › skips /admin 
@astrojs/prefetch:test: 
@astrojs/prefetch:test:     browserType.launch: Chromium distribution 'chrome' is not found at /opt/google/chrome/chrome
@astrojs/prefetch:test:     Run "npx playwright install chrome"
@astrojs/prefetch:test: 
@astrojs/prefetch:test:         at Object._baseTest.extend.browser.scope [as fn] (/astro/node_modules/.pnpm/@playwright+test@1.25.1/node_modules/@playwright/test/lib/index.js:222:51)
@astrojs/prefetch:test:         at /astro/node_modules/.pnpm/@playwright+test@1.25.1/node_modules/@playwright/test/lib/fixtures.js:112:81
@astrojs/prefetch:test: 
@astrojs/prefetch:test:   2) [Chrome Stable] › test/basic-prefetch.test.js:55:4 › Basic prefetch › build › prefetches rel="prefetch" links › skips /admin 
@astrojs/prefetch:test: 
@astrojs/prefetch:test:     TypeError: process.stdout.clearLine is not a function
@astrojs/prefetch:test: 
@astrojs/prefetch:test:       43 |
@astrojs/prefetch:test:       44 | 		test.beforeAll(async ({ astro }) => {
@astrojs/prefetch:test:     > 45 | 			await astro.build();
@astrojs/prefetch:test:          | 			^
@astrojs/prefetch:test:       46 | 			previewServer = await astro.preview();
@astrojs/prefetch:test:       47 | 		});
@astrojs/prefetch:test:       48 |
@astrojs/prefetch:test: 
@astrojs/prefetch:test:         at Object.buildEnd (file:///astro/node_modules/.pnpm/vite@3.0.9_sass@1.54.5/node_modules/vite/dist/node/chunks/dep-0fc8e132.js:12619:36)
@astrojs/prefetch:test:         at file:///astro/node_modules/.pnpm/rollup@2.77.3/node_modules/rollup/dist/es/shared/rollup.js:22695:37
@astrojs/prefetch:test:         at async Promise.all (index 3)
@astrojs/prefetch:test:         at file:///astro/node_modules/.pnpm/rollup@2.77.3/node_modules/rollup/dist/es/shared/rollup.js:23649:13
@astrojs/prefetch:test:         at catchUnfinishedHookActions (file:///astro/node_modules/.pnpm/rollup@2.77.3/node_modules/rollup/dist/es/shared/rollup.js:23041:20)
@astrojs/prefetch:test:         at rollupInternal (file:///astro/node_modules/.pnpm/rollup@2.77.3/node_modules/rollup/dist/es/shared/rollup.js:23639:5)
@astrojs/prefetch:test:         at doBuild (file:///astro/node_modules/.pnpm/vite@3.0.9_sass@1.54.5/node_modules/vite/dist/node/chunks/dep-0fc8e132.js:43620:24)
@astrojs/prefetch:test:         at Module.build (file:///astro/node_modules/.pnpm/vite@3.0.9_sass@1.54.5/node_modules/vite/dist/node/chunks/dep-0fc8e132.js:43468:16)
@astrojs/prefetch:test:         at clientBuild (file:///astro/packages/astro/dist/core/build/static-build.js:195:23)
@astrojs/prefetch:test:         at staticBuild (file:///astro/packages/astro/dist/core/build/static-build.js:64:3)
@astrojs/prefetch:test:         at AstroBuilder.build (file:///astro/packages/astro/dist/core/build/index.js:83:5)
@astrojs/prefetch:test:         at AstroBuilder.run (file:///astro/packages/astro/dist/core/build/index.js:124:7)
@astrojs/prefetch:test:         at build (file:///astro/packages/astro/dist/core/build/index.js:22:3)
@astrojs/prefetch:test:         at file:///astro/packages/integrations/prefetch/test/basic-prefetch.test.js:45:4
@astrojs/prefetch:test: 
@astrojs/prefetch:test:   3) [Chrome Stable] › test/custom-selectors.test.js:27:4 › Custom prefetch selectors › dev › prefetches links by custom selector › only prefetches /contact 
@astrojs/prefetch:test: 
@astrojs/prefetch:test:     browserType.launch: Chromium distribution 'chrome' is not found at /opt/google/chrome/chrome
@astrojs/prefetch:test:     Run "npx playwright install chrome"
@astrojs/prefetch:test: 
@astrojs/prefetch:test:         at Object._baseTest.extend.browser.scope [as fn] (/astro/node_modules/.pnpm/@playwright+test@1.25.1/node_modules/@playwright/test/lib/index.js:222:51)
@astrojs/prefetch:test:         at /astro/node_modules/.pnpm/@playwright+test@1.25.1/node_modules/@playwright/test/lib/fixtures.js:112:81
@astrojs/prefetch:test: 
@astrojs/prefetch:test:   4) [Chrome Stable] › test/custom-selectors.test.js:60:4 › Custom prefetch selectors › build › prefetches links by custom selector › only prefetches /contact 
@astrojs/prefetch:test: 
@astrojs/prefetch:test:     TypeError: process.stdout.clearLine is not a function
@astrojs/prefetch:test: 
@astrojs/prefetch:test:       48 |
@astrojs/prefetch:test:       49 | 		test.beforeAll(async ({ astro }) => {
@astrojs/prefetch:test:     > 50 | 			await astro.build();
@astrojs/prefetch:test:          | 			^
@astrojs/prefetch:test:       51 | 			previewServer = await astro.preview();
@astrojs/prefetch:test:       52 | 		});
@astrojs/prefetch:test:       53 |
@astrojs/prefetch:test: 
@astrojs/prefetch:test:         at Object.buildEnd (file:///astro/node_modules/.pnpm/vite@3.0.9_sass@1.54.5/node_modules/vite/dist/node/chunks/dep-0fc8e132.js:12619:36)
@astrojs/prefetch:test:         at file:///astro/node_modules/.pnpm/rollup@2.77.3/node_modules/rollup/dist/es/shared/rollup.js:22695:37
@astrojs/prefetch:test:         at async Promise.all (index 3)
@astrojs/prefetch:test:         at file:///astro/node_modules/.pnpm/rollup@2.77.3/node_modules/rollup/dist/es/shared/rollup.js:23649:13
@astrojs/prefetch:test:         at catchUnfinishedHookActions (file:///astro/node_modules/.pnpm/rollup@2.77.3/node_modules/rollup/dist/es/shared/rollup.js:23041:20)
@astrojs/prefetch:test:         at rollupInternal (file:///astro/node_modules/.pnpm/rollup@2.77.3/node_modules/rollup/dist/es/shared/rollup.js:23639:5)
@astrojs/prefetch:test:         at doBuild (file:///astro/node_modules/.pnpm/vite@3.0.9_sass@1.54.5/node_modules/vite/dist/node/chunks/dep-0fc8e132.js:43620:24)
@astrojs/prefetch:test:         at Module.build (file:///astro/node_modules/.pnpm/vite@3.0.9_sass@1.54.5/node_modules/vite/dist/node/chunks/dep-0fc8e132.js:43468:16)
@astrojs/prefetch:test:         at clientBuild (file:///astro/packages/astro/dist/core/build/static-build.js:195:23)
@astrojs/prefetch:test:         at staticBuild (file:///astro/packages/astro/dist/core/build/static-build.js:64:3)
@astrojs/prefetch:test:         at AstroBuilder.build (file:///astro/packages/astro/dist/core/build/index.js:83:5)
@astrojs/prefetch:test:         at AstroBuilder.run (file:///astro/packages/astro/dist/core/build/index.js:124:7)
@astrojs/prefetch:test:         at build (file:///astro/packages/astro/dist/core/build/index.js:22:3)
@astrojs/prefetch:test:         at file:///astro/packages/integrations/prefetch/test/custom-selectors.test.js:50:4
@astrojs/prefetch:test: 
@astrojs/prefetch:test: 
@astrojs/prefetch:test:   4 failed
@astrojs/prefetch:test:     [Chrome Stable] › test/basic-prefetch.test.js:19:4 › Basic prefetch › dev › prefetches rel="prefetch" links › skips /admin 
@astrojs/prefetch:test:     [Chrome Stable] › test/basic-prefetch.test.js:55:4 › Basic prefetch › build › prefetches rel="prefetch" links › skips /admin 
@astrojs/prefetch:test:     [Chrome Stable] › test/custom-selectors.test.js:27:4 › Custom prefetch selectors › dev › prefetches links by custom selector › only prefetches /contact 
@astrojs/prefetch:test:     [Chrome Stable] › test/custom-selectors.test.js:60:4 › Custom prefetch selectors › build › prefetches links by custom selector › only prefetches /contact 
@astrojs/prefetch:test:  ELIFECYCLE  Test failed. See above for more details.
@astrojs/prefetch:test: Error: command finished with error: command (packages/integrations/prefetch) pnpm run test exited (1)

```

It's hard to write proper tests when you can only run them in Github Actions

File: packages/integrations/prefetch/src/client.ts [hunk 1]
```diff
@@ -53,7 +54,14 @@ async function preloadHref(link: HTMLAnchorElement) {
 		const html = parser.parseFromString(contents, 'text/html');
 		const styles = Array.from(html.querySelectorAll<HTMLLinkElement>('link[rel="stylesheet"]'));
 
-		await Promise.all(styles.map((el) => fetch(el.href)));
+		await Promise.all(
+			styles
+				.filter((el) => !loadedStyles.has(el.href))
+				.map((el) => {
+					loadedStyles.add(el.href);
+					return fetch(el.href);
```
Comment (natemoo-re) [inline packages/integrations/prefetch/src/client.ts:62]: This promise needed to be returned, which might fix the tests! Watching CI to see.

Comment (matthewp) [reply packages/integrations/prefetch/src/client.ts:62]: @Jelenkee can you try this?

**Comment by matthewp** [comment]
Closing because the tests are failing. @Jelenkee If you get time, the above suggestion might fix the tests, in which case you can resubmit a new PR.

**Pull request closed without merging**